crates/doc/src/inference.rs

```diff
@@ -444,6 +460,84 @@ impl ObjShape {
             None
         }
     }
+
+    fn widen<'n, N>(&mut self, fields: &'n N::Fields, loc: Location, is_first_time: bool) -> bool
+    where
+        N: AsNode,
+    {
+        use crate::{Field, Fields};
+        // If a particular location defines `additionalProperties` as a subschema
```

Comment:
~nit:~ this comment could be a bit crisper. For example:

```
Two possible cases:
* This schema has individual `properties` as well as `additionalProperties: false`,
   meaning no other properties are possible. We recursively widen each of the input
  `fields` into their respective `properties`, adding new ones as required.
* This schema has `additionalProperties` _other_ than `false`.
   In this case, we apply each of `fields` to widen matched `properties`,
   and otherwise widen `additionalProperties` where not matched.
```

Right, and this is the subtle power of good comments. Writing this out made me realize that our behavior is incorrect in the general case, where we're widening an `ObjShape` that may have started as an actual schema.

To be fully correct, we _would_ need to look for matched `properties` and even `patternProperties` to widen _before_ we resort to widening `additionalProperties`.

In your shoes I would correct the implementation now -- because I'll otherwise forget about it -- but I'm okay punting given the time pressure for this feature. If we do punt, please add a good comment here and issue on this defect.

Reply:
Fixed and tested that:
* We widen explicitly named properties even if `additionalProperties` is non-false
* We widen matching `patternProperties` before widening `additionalProperties`